libnopegl/src/backends/gl/pipeline_gl.c

```diff
@@ -225,23 +227,10 @@ static int build_texture_bindings(struct pipeline *s)
             return NGL_ERROR_MEMORY;
     }
 
-    if (nb_textures > limits->max_texture_image_units) {
-        LOG(ERROR, "number of texture units (%zu) exceeds device limits (%u)", nb_textures, limits->max_texture_image_units);
-        return NGL_ERROR_GRAPHICS_LIMIT_EXCEEDED;
-    }
-
-    const struct gpu_ctx *gpu_ctx = s->gpu_ctx;
-    const struct gpu_ctx_gl *gpu_ctx_gl = (struct gpu_ctx_gl *)gpu_ctx;
-    const struct glcontext *gl = gpu_ctx_gl->glcontext;
-    if (nb_images && !(gl->features & NGLI_FEATURE_GL_SHADER_IMAGE_LOAD_STORE)) {
-        LOG(ERROR, "context does not support shader image load store operations");
-        return NGL_ERROR_GRAPHICS_UNSUPPORTED;
-    }
-
-    if (nb_images > limits->max_image_units) {
-        LOG(ERROR, "number of image units (%zu) exceeds device limits (%u)", nb_images, limits->max_image_units);
-        return NGL_ERROR_GRAPHICS_LIMIT_EXCEEDED;
-    }
+    ngli_assert(nb_textures <= limits->max_texture_image_units);
+    if (nb_images)
+        ngli_assert(gl->features & NGLI_FEATURE_GL_SHADER_IMAGE_LOAD_STORE);
+    ngli_assert(nb_images <= limits->max_image_units);
```

Comment:
Where do you we have clean user checks for these?

Reply:
Added a check in pass.c in a separate commit for image load store operations support and I dropped the ngli_assert() on nb_textures / nb_images. This would require adding checks in before calling pipeline_init().